libraries/AP_Proximity/AP_Proximity_Boundary_3D.h

```diff
@@ -0,0 +1,147 @@
+    // This distance can then be used for Obstacle Avoidance
+    // Assume detected obstacle is horizontal (zero pitch), if no pitch is passed
+    void set_face_attributes(const Face &face, float pitch, float yaw, float distance);
+    void set_face_attributes(const Face &face, float yaw, float distance) { set_face_attributes(face, 0, yaw, distance); }
+
+    // add a distance to the boundary if it is shorter than any other provided distance since the last time the boundary was reset
+    // pitch and yaw are in degrees, distance is in meters
+    void add_distance(float pitch, float yaw, float distance);
+    void add_distance(float yaw, float distance) { add_distance(0, yaw, distance); }
+
+    // update boundary points used for simple avoidance based on a single sector and pitch distance changing
+    //   the boundary points lie on the line between sectors meaning two boundary points may be updated based on a single sector's distance changing
+    //   the boundary point is set to the shortest distance found in the two adjacent sectors, this is a conservative boundary around the vehicle
+    void update_boundary(const Face &face);
+
+    // update middle layer boundary points
+    void update_middle_boundary();
+
+    // reset boundary.  marks all distances as invalid
+    void reset();
+
+    // Reset this location, specified by Face object, back to default
+    // i.e Distance is marked as not-valid
+    void reset_face(const Face &face);
+
+    // get distance for a face.  returns true on success and fills in distance argument with distance in meters
+    bool get_distance(const Face &face, float &distance) const;
+
+    // Get the total number of obstacles 
+    uint8_t get_obstacle_count() const;
+
+    // Appropriate layer and sector are found from the passed obstacle_num
+    // This function then draws a line between this sector, and sector + 1 at the given layer
+    // Then returns the closest point on this line from vehicle, in body-frame. 
+    // False is returned if the obstacle_num provided does not produce a valid obstacle 
+    bool get_obstacle(uint8_t obstacle_num, Vector3f& vec_to_boundary) const;
+
+    // Appropriate layer and sector are found from the passed obstacle_num
+    // This function then draws a line between this sector, and sector + 1 at the given layer
+    // Then returns the closest point on this line from the segment that was passed, in body-frame.
+    // Used by GPS based Simple Avoidance  - for "brake mode" 
+    // FLT_MAX is returned if the obstacle_num provided does not produce a valid obstacle
+    float distance_to_obstacle(uint8_t obstacle_num, const Vector3f& seg_start, const Vector3f& seg_end, Vector3f& closest_point) const;
+
+    // get distance and angle to closest object (used for pre-arm check)
+    //   returns true on success, false if no valid readings
+    bool get_closest_object(float& angle_deg, float &distance) const;
```

Comment:
are "objects" and "obstacles" different?  from a layman's point of view they are synonyms so if they are the same to us as well we should consolidate on a single word (I don't mind which).  I suspect they are calculated differently (one is a point that a sensor has actually seen while another is an interpolated point taken from the boundary)... in which case it is OK to leave them named like this and we will fix it later.

Reply:
Yes you're right, they are calculated differently...I have used obstacle where GPS based Simple Avoidance is accessing the 3-D boundary. Object (From current master, i.e unchanged by this PR), is mostly for non-gps Simple Avoidance which uses the sensor values directly, and doesn't really use the 3-D boundary.